firedrake/projection.py

```diff
@@ -97,3 +97,41 @@ def project(v, V, bcs=None, mesh=None,
            solver_parameters=solver_parameters,
            form_compiler_parameters=form_compiler_parameters)
     return ret
+
+
+class Projector(object):
+    """
+    A projector projects a UFL expression into a function space
+    and places the result in a function from that function space,
+    allowing the solver to be reused.
+
+    :arg v: the :class:`.Expression`, :class:`ufl.Expr` or
+         :class:`.Function` to project
+    :arg v_out: :class:`.Function` to put the result in
+    :arg solver_parameters: parameters to pass to the solver used when
+         projecting.    
+    """
+
+    def __init__(self, v, v_out, solver_parameters = None):
+        
+        V = v_out.function_space
+        
+        p = TestFunction(V)
+        q = TrialFunction(V)
+        
+        a = inner(p, q)*dx
+        L = inner(p, v)*dx
+
+        problem = LinearVariationalProblem(a, L, v_out)
+        
+        if(solver_parameters == None):
+            self.solver = LinearVariationalSolver(problem)
+        else:
+            self.solver = LinearVariationalSolver(problem, solver_parameters)
```

Comment:
Probably won't be a problem if I don't support Expression actually, just
ufl.Expr.

--cjc

On 29 January 2016 at 10:20, Lawrence Mitchell <notifications@github.com>
wrote:

> In firedrake/projection.py
> <https://github.com/firedrakeproject/firedrake/pull/686#discussion_r51245105>
> :
>
> > +    def __init__(self, v, v_out, solver_parameters = None):
> > +
> > +        V = v_out.function_space
> > +
> > +        p = TestFunction(V)
> > +        q = TrialFunction(V)
> > +
> > +        a = inner(p, q)*dx
> > +        L = inner(p, v)*dx
> > +
> > +        problem = LinearVariationalProblem(a, L, v_out)
> > +
> > +        if(solver_parameters == None):
> > +            self.solver = LinearVariationalSolver(problem)
> > +        else:
> > +            self.solver = LinearVariationalSolver(problem, solver_parameters)
>
> Although having said that, you probably want to set the default ksp_type,
> to cg
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/firedrakeproject/firedrake/pull/686/files#r51245105>.
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916